openmdao/visualization/opt_report/opt_report.py

```diff
@@ -0,0 +1,1037 @@
+
+# variable value formatting
+_variable_label_format = '5.3g'
+
+# https://jfly.uni-koeln.de/color/
+# scalar visual parameters
+_plot_x_max = 1.0
+_plot_y_max = 1.0
+_plot_y_margin = 0.2
+_lower_plot = _plot_x_max / 3.
+_upper_plot = 2 * _plot_x_max / 3.
+_pointer_half_width = 0.04
+_pointer_height = 0.4
+_text_height = 0.3
+_pointer_line_width = 0.05
+_near_bound_highlight_half_width = 0.05
+_near_bound_highlight_half_width_y_min = 0.0
+_near_bound_highlight_half_width_y_max = _plot_y_max
+_near_bound_highlight_alpha = 0.7
+_equality_bound_width = 0.01
+# For the ellipsis that are displayed when scalar is way outside of bounds
+_ellipsis_x_offset = 0.2
+_ellipse_width = 0.01
+_ellipse_height = 0.15
+
+
+def _optimizer_report_register():
+    register_report('optimizer', opt_report, 'Summary of optimization', 'Driver', 'run', 'post')
+
+
+def opt_report(driver, outfile=None):
+    """
+    Write a summary of the optimization to the given file.
+
+    Parameters
+    ----------
+    driver : Driver
+        The Driver for which the optimization report is being generated.
+    outfile : str or None
+        The path to the HTML file to be written.  If None (default), write to the default report
+        output path.
+    """
+    prob = driver._problem()
+    if not driver.supports['optimization']:
+        driver_class = type(driver).__name__
+        issue_warning(f"The optimizer report is not applicable for the {driver_class} Driver "
+                      "which does not support optimization", category=DriverWarning)
+        return
+
+    # only create report on rank 0
+    if MPI and MPI.COMM_WORLD.rank != 0:
+        return
+
+    if not outfile:
+        outfile = _default_optimizer_report_filename
+
+    outfilepath = str(pathlib.Path(prob.get_reports_dir()).joinpath(outfile))
+
+    print(f"in opt_report, outfilepath = {outfilepath}")
+    print(f"in opt_report, os.getcwd() = {os.getcwd()}")
+
+    if not tabulate:
+        s = '''
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <title>Optimization Report - Unable to Generate</title>
+</head>
+<body>
+
+<p>The optimization report requires the tabulate package but it is not currently installed.</p>
+
+<p>To install the tablulate module:</p>
+
+<code>
+pip install tabulate
+</code>
+
+</body>
+</html>
+        '''
+        with open(outfilepath, 'w') as f:
+            f.write(s)
+
+        return
+
+    driver_scaling = True
+
+    # Collect data from the problem
+    abs2prom = prob.model._var_abs2prom
+
+    # Collect the entire array of array valued desvars and constraints (ignore indices)
+    objs_vals = {}
+    desvars_vals = {}
+    cons_vals = {}
+
+    objs_meta = _prom_name_dict(prob.driver._objs, abs2prom)
+    desvars_meta = _prom_name_dict(prob.driver._designvars, abs2prom)
+    cons_meta = _prom_name_dict(prob.driver._cons, abs2prom)
+
+    with prob.model._scaled_context_all():
+        for abs_name, meta in driver._objs.items():
```

Comment:
Couldn't you just use the promoted name dicts you just created above and loop over those instead of looping over _objs, _designvars, and _cons and repeating the conversions from abs to promoted names in this and the next couple of loops?